### PR TITLE
274: Fixed rendering issues

### LINKED ISSUE #274: Pixel-perfect rendering
The library has a lot of small rendering issues, generally on HiDPI devices. We have to migrate to [fancy canvas library](https://github.com/tradingview/fancy-canvas) in order of fixing it.

### PR DESCRIPTION
**Pull request by eugene-korobko**
**Type of PR:** bugfix, enhancement

**PR checklist:**

- [x] Addresses an existing issue: fixes #274
- [x] Includes tests
- [ ] Documentation update

### CODE REVIEW

File: src/api/options/time-scale-options-defaults.ts [hunk 1]
```diff
@@ -11,4 +11,5 @@ export const timeScaleOptionsDefaults: TimeScaleOptions = {
 	visible: true,
 	timeVisible: false,
 	secondsVisible: true,
+	smoothZoom: true,
```
Comment (timocov) [inline src/api/options/time-scale-options-defaults.ts]: I guess we need to add it to the docs as well?

File: src/renderers/iprice-axis-view-renderer.ts [hunk 2]
```diff
@@ -17,7 +17,7 @@ export interface PriceAxisViewRendererData {
 	tickVisible: boolean;
 	borderVisible: boolean;
 	lineWidth?: LineWidth;
-	lineStyle: LineStyle;
+	offset: number;
```
Comment (timocov) [inline src/renderers/iprice-axis-view-renderer.ts]: `lineStyle` isn't used anywhere? What is `offset`?

File: src/renderers/histogram-renderer.ts [hunk 3]
```diff
@@ -1,50 +1,136 @@
...
+				const width = current.right - current.left + 1;
+				if (width > minWidth) {
+					if (current.roundedCenter > current.center) {
+						current.right -= 1;
+					} else {
+						current.left += 1;
+					}
+				}
+			}
 		}
```
Comment (timocov) [inline src/renderers/histogram-renderer.ts]:
```suggestion
```

File: src/renderers/grid-renderer.ts [hunk 4]
```diff
@@ -60,13 +65,18 @@ export class GridRenderer implements IPaneRenderer {
...
 
 			ctx.stroke();
 		}
 
+		if (lineWidth % 2) {
+			ctx.translate(-0.5, -0.5);
+		}
+
 		ctx.restore();
+
```
Comment (timocov) [inline src/renderers/grid-renderer.ts]:
```suggestion
```

File: src/renderers/grid-renderer.ts [hunk 5]
```diff
@@ -60,13 +65,18 @@ export class GridRenderer implements IPaneRenderer {
...
-				ctx.lineTo(width, y);
+				ctx.moveTo(-lineWidth, y);
+				ctx.lineTo(width + lineWidth, y);
 			}
 
 			ctx.stroke();
 		}
 
+		if (lineWidth % 2) {
+			ctx.translate(-0.5, -0.5);
```
Comment (timocov) [inline src/renderers/grid-renderer.ts]: should we do that in case of `ctx.restore()` call below?

File: src/renderers/histogram-renderer.ts [hunk 6]
```diff
@@ -1,50 +1,136 @@
...
-			ctx.rect(left, y, right - left, histogramBase - y);
+			ctx.fillStyle = item.color;
+			ctx.fillRect(current.left, y, current.right - current.left + 1, histogramBase - y + lineWidth);
+		}
+	}
+
+	// tslint:disable-next-line: cyclomatic-complexity
+	private _fillPrecalculatedCache(pixelRatio: number): void {
+		if (this._data === null || this._data.items.length === 0 || this._data.visibleRange === null) {
+			return;
```
Comment (timocov) [inline src/renderers/histogram-renderer.ts]: should we clear cache `this._precalculatedCache` in this case? if so, it'd be great to see test case for that if it's possible.

File: src/api/options/time-scale-options-defaults.ts (hunk 1 above)
Comment (timocov) [reply src/api/options/time-scale-options-defaults.ts]: also, if it's possible, add tests for that

File: src/helpers/mathex.ts [hunk 7]
```diff
@@ -46,3 +46,7 @@ export function min(arr: number[]): number {
 
 	return minVal;
 }
+
+export function even(x: number): number {
```
Comment (Nipheris) [inline src/helpers/mathex.ts]: `ceilingEven`? :)

File: src/model/time-scale.ts [hunk 8]
```diff
@@ -55,6 +55,7 @@ export interface TimeScaleOptions {
 	visible: boolean;
 	timeVisible: boolean;
 	secondsVisible: boolean;
+	smoothZoom: boolean;
```
Comment (Nipheris) [inline src/model/time-scale.ts]: Should be documented too?

File: src/renderers/bars-renderer.ts [hunk 9]
```diff
@@ -26,15 +26,25 @@ export class PaneRendererBars implements IPaneRenderer {
...
 			return;
 		}
 
+		this._barWidth = Math.max(1, Math.floor(optimalBarWidth(this._data.barSpacing, pixelRatio)));
+
+		// if we have enough pixels between candles
+		if (this._barWidth >= 2) {
+			const lineWidth = Math.floor(pixelRatio);
+			if ((lineWidth % 2) !== (this._barWidth % 2)) {
+				this._barWidth--;
```
Comment (Nipheris) [inline src/renderers/bars-renderer.ts]: Maybe we can comment such +1/-1 alignments?

File: src/renderers/candlesticks-renderer.ts [hunk 10]
```diff
@@ -29,18 +29,29 @@ const enum Constants {
...
 		if (this._data === null || this._data.bars.length === 0 || this._data.visibleRange === null) {
 			return;
 		}
 
+		// now we know pixelRatio and we could calculate barWidth effectively
+		this._barWidth = optimalCandlestickWidth(this._data.barSpacing, pixelRatio);
+		// if we have enough pixels between candles
+		if (this._barWidth >= 2) {
+			const wickWidth = Math.floor(pixelRatio);
+			if ((wickWidth % 2) !== (this._barWidth % 2)) {
```
Comment (Nipheris) [inline src/renderers/candlesticks-renderer.ts]: The same non-trivial logic. Add some comments?

File: src/api/options/time-scale-options-defaults.ts (hunk 1 above)
Comment (eugene-korobko) [reply src/api/options/time-scale-options-defaults.ts]: we decided not to implement this, so, removed

File: src/renderers/iprice-axis-view-renderer.ts (hunk 2 above)
Comment (eugene-korobko) [reply src/renderers/iprice-axis-view-renderer.ts]: 1. Yes, `lineStyle` was never used actually.
2. Offset is an offset from the edge. We need this to separane pane part of label from price axis part

File: src/renderers/grid-renderer.ts [hunk 11]
```diff
@@ -33,25 +33,30 @@ export class GridRenderer implements IPaneRenderer {
...
+		ctx.lineWidth = lineWidth;
 
 		const height = Math.ceil(this._data.h * pixelRatio);
 		const width = Math.ceil(this._data.w * pixelRatio);
 
 		ctx.save();
-		ctx.translate(0.5, 0.5);
+
+		if (lineWidth % 2) {
+			ctx.translate(0.5, 0.5);
```
Comment (Nipheris) [inline src/renderers/grid-renderer.ts]: Maybe wrap all the logic that depends on this correction in some helper, similar to `drawScaled`? Something like `strokeInPixel`?)

File: src/renderers/grid-renderer.ts (hunk 5 above)
Comment (eugene-korobko) [reply src/renderers/grid-renderer.ts]: no, removed

File: src/renderers/iprice-axis-view-renderer.ts (hunk 2 above)
Comment (timocov) [reply src/renderers/iprice-axis-view-renderer.ts]: > We need this to separane pane part of label from price axis part

Not sure that understand that. Maybe we need to remove it somehow?

File: src/renderers/histogram-renderer.ts (hunk 6 above)
Comment (eugene-korobko) [reply src/renderers/histogram-renderer.ts]: I'm not sure if this case is possible, however clearing looks reasonable here.

File: src/model/time-scale.ts (hunk 8 above)
Comment (eugene-korobko) [reply src/model/time-scale.ts]: already removed

File: src/renderers/bars-renderer.ts (hunk 9 above)
Comment (eugene-korobko) [reply src/renderers/bars-renderer.ts]: added

File: src/renderers/grid-renderer.ts (hunk 11 above)
Comment (eugene-korobko) [reply src/renderers/grid-renderer.ts]: hm, ok, looks to be ok

File: src/renderers/iprice-axis-view-renderer.ts (hunk 2 above)
Comment (timocov) [reply src/renderers/iprice-axis-view-renderer.ts]: @Nipheris thoughts?

**Pull request merged**